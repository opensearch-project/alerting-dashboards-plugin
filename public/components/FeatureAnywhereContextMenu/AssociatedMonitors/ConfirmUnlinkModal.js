/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';

export const ConfirmUnlinkDetectorModal = (props) => {
  const [isModalLoading, setIsModalLoading] = useState(false);
  const isLoading = isModalLoading || props.isListLoading;
  return (
    <EuiOverlayMask>
      <EuiModal data-test-subj="startDetectorsModal" onClose={props.onHide} maxWidth={450}>
        <EuiModalHeader>
          <EuiModalHeaderTitle>{'Remove association?'}&nbsp;</EuiModalHeaderTitle>
        </EuiModalHeader>
        <EuiModalBody>
          <EuiText>
            Removing association unlinks {props.monitor.name} monitor from the visualization but
            does not delete it. The monitor association can be restored.
          </EuiText>
          <EuiSpacer size="s" />
        </EuiModalBody>
        <EuiModalFooter>
          {isLoading ? null : (
            <EuiButtonEmpty data-test-subj="cancelButton" onClick={props.onHide}>
              Cancel
            </EuiButtonEmpty>
          )}
          <EuiButton
            data-test-subj="confirmButton"
            color="primary"
            fill
            isLoading={isLoading}
            onClick={async () => {
              setIsModalLoading(true);
              props.onConfirm();
            }}
          >
            {'Remove association'}
          </EuiButton>
        </EuiModalFooter>
      </EuiModal>
    </EuiOverlayMask>
  );
};
