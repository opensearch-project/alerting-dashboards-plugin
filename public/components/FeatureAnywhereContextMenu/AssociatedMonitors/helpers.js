/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useMemo } from 'react';
import { EuiHealth, EuiLink } from '@elastic/eui';
import { stateToLabel } from '../../../utils/contextMenu/monitors';
import { dateOptionsLong } from '../../../utils/contextMenu/helpers';
import { PLUGIN_NAME } from '../../../../utils/constants';

export const useColumns = ({ onUnlink, onEdit }) => {
  const columns = useMemo(
    () => [
      {
        field: 'name',
        name: 'Monitor Name',
        sortable: true,
        truncateText: true,
        width: '50%',
        render: (name, monitor) => (
          <EuiLink href={`${PLUGIN_NAME}#/monitors/${monitor.id}`} target="_blank">
            {name}
          </EuiLink>
        ),
      },
      {
        field: 'state',
        name: 'State',
        sortable: true,
        width: '105px',
        render: (state) => (
          <EuiHealth color={stateToLabel[state].color}>{stateToLabel[state].label}</EuiHealth>
        ),
      },
      {
        field: 'date',
        name: 'Latest alert',
        sortable: true,
        truncateText: true,
        width: '50%',
        render: (date) => {
          if (date) {
            return new Intl.DateTimeFormat('default', dateOptionsLong).format(date);
          } else {
            return '-';
          }
        },
      },
      {
        name: 'Actions',
        actions: [
          {
            type: 'icon',
            name: 'Edit monitor',
            description: 'Edit monitor',
            icon: 'pencil',
            onClick: onEdit,
          },
          {
            type: 'icon',
            name: 'Unlink monitor',
            description: 'Unlink monitor',
            icon: 'unlink',
            onClick: onUnlink,
          },
        ],
      },
    ],
    [onUnlink, onEdit]
  );
  return columns;
};
