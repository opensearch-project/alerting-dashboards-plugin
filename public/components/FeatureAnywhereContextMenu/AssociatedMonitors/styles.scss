/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

@import '@elastic/eui/src/global_styling/variables/index';

.associated-monitors {
  height: 100%;
  display: flex;
  flex-direction: column;

  .euiFlyoutBody__overflowContent {
    height: 100%;
    padding-bottom: 0;
  }

  &__flex-group {
    height: 100%;
  }

  &__associate-button {
    flex: 0 0 auto;
  }
}
