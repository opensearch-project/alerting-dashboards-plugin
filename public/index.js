/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AlertingPlugin } from './plugin';

export function plugin(initializerContext) {
  return new AlertingPlugin(initializerContext);
}
