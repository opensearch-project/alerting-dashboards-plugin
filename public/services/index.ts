/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import NotificationService from './NotificationService';

export * from './services';
export { NotificationService };
