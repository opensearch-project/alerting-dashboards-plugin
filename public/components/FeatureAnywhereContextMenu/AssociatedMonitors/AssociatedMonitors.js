import React, { useCallback, useMemo, useState } from 'react';
import {
  EuiFlyoutHeader,
  EuiTitle,
  EuiText,
  EuiSpacer,
  EuiInMemoryTable,
  EuiFlyoutBody,
  EuiEmptyPrompt,
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiLoadingSpinner,
  EuiTextColor,
} from '@elastic/eui';
import { get } from 'lodash';
import './styles.scss';
import { useColumns } from './helpers';
import { getSavedAugmentVisLoader } from '../../../../../../src/plugins/vis_augmenter/public';
import { ConfirmUnlinkDetectorModal } from './ConfirmUnlinkModal';

const AssociatedMonitors = ({ embeddable, closeFlyout, setFlyoutMode, monitors }) => {
  const title = embeddable.getTitle();
  const [modalState, setModalState] = useState(undefined);
  const onUnlink = useCallback((item) => {
    setModalState({
      detector: {
        name: item.name,
        id: item.id,
      },
    });
  }, []);
  const onEdit = useCallback(
    (item) => {
      window.open(`alerting#/monitors/${item.id}?action=update-monitor`, '_blank');
      // closeFlyout();
    },
    [closeFlyout]
  );
  const savedObjectLoader = useMemo(() => getSavedAugmentVisLoader(), []);
  const onUnlinkDetector = useCallback(async () => {
    // setIsLoadingFinalDetectors(true);
    await savedObjectLoader.findAll().then(async (resp) => {
      if (resp != undefined) {
        const savedAugmentObjects = get(resp, 'hits', []);
        // gets all the saved object for this visualization
        const savedAugmentForThisVisualization = savedAugmentObjects.filter(
          (savedObj) => get(savedObj, 'visId', '') === embeddable.vis.id
        );

        // find saved Augment object matching detector we want to unlink
        // There should only be one detector and vis pairing
        const savedAugmentToUnlink = savedAugmentForThisVisualization.find(
          (savedObject) => get(savedObject, 'pluginResourceId', '') === modalState.detector.id
        );
        const savedObjectToUnlinkId = get(savedAugmentToUnlink, 'id', '');
        await savedObjectLoader
          .delete(savedObjectToUnlinkId)
          .catch((error) => {
            // core.notifications.toasts.addDanger(
            //   prettifyErrorMessage(
            //     `Error unlinking selected detector: ${error}`
            //   )
            // );
            console.log(`Error unlinking selected detector: ${error}`);
          })
          .finally(() => {
            // getDetectors();
            handleHideModal();
            closeFlyout();
          });
      }
    });
  }, [closeFlyout, savedObjectLoader, modalState]);
  const columns = useColumns({ onUnlink, onEdit });
  const emptyMsg = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body={`There are no alerting monitors associated with ${title} visualization.`}
    />
  );
  const noResultsMsg = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body="There are no alerting monitors that match the search result."
    />
  );
  const loadingMsg = <EuiEmptyPrompt body={<EuiLoadingSpinner size="l" />} />;
  const tableProps = {
    items: monitors || [],
    columns,
    search: {
      box: {
        disabled: !monitors || monitors.length === 0,
        incremental: true,
        schema: true,
      },
    },
    hasActions: true,
    pagination: true,
    sorting: true,
    message: monitors ? (monitors.length ? noResultsMsg : emptyMsg) : loadingMsg,
    loading: !monitors,
  };

  const handleHideModal = useCallback(() => {
    setModalState(undefined);
  }, []);

  return (
    <div className="associated-monitors">
      {modalState ? (
        <ConfirmUnlinkDetectorModal
          detector={modalState.detector}
          onHide={handleHideModal}
          onConfirm={onUnlinkDetector}
          isListLoading={!monitors}
        />
      ) : null}
      <EuiFlyoutHeader hasBorder>
        <EuiTitle>
          <h2 id="associated-monitors__title">
            Associated monitors{' '}
            {monitors && (
              <EuiTextColor color={monitors.length > 0 ? 'default' : 'subdued'}>
                ({monitors.length})
              </EuiTextColor>
            )}
          </h2>
        </EuiTitle>
      </EuiFlyoutHeader>
      <EuiFlyoutBody>
        <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
          <EuiFlexItem>
            <EuiText>
              <h4>{title}</h4>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButton iconType="link" fill onClick={() => setFlyoutMode('existing')}>
              Associate a monitor
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer size="l" />
        <EuiInMemoryTable {...tableProps} />
      </EuiFlyoutBody>
    </div>
  );
};

export default AssociatedMonitors;
