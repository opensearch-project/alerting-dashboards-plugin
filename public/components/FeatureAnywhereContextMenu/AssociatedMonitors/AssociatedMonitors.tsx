/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useState } from 'react';
import {
  EuiButton,
  EuiEmptyPrompt,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyoutBody,
  EuiFlyoutHeader,
  EuiInMemoryTable,
  EuiLoadingSpinner,
  EuiSpacer,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import './styles.scss';
import { useColumns } from './helpers';
import { ConfirmUnlinkDetectorModal } from './ConfirmUnlinkModal';
import { deleteAlertingAugmentVisSavedObj } from '../../../utils/savedObjectHelper';
import { getNotifications } from '../../../services';

const AssociatedMonitors = ({ embeddable, closeFlyout, setFlyoutMode, monitors, isAssociateAllowed, limitReachedCallout, setAssociatedMonitors }) => {
  const title = embeddable.vis.title;
  const [modalState, setModalState] = useState(undefined);
  const notifications = getNotifications();
  const onUnlink = useCallback((item) => {
    setModalState({
      monitor: {
        name: item.name,
        id: item.id,
      },
    });
  }, []);
  const onEdit = useCallback(
    (item) => {
      window.open(`alerting#/monitors/${item.id}?action=update-monitor`, '_blank');
    },
    []
  );
  const onUnlinkMonitor = useCallback(async () => {
    try {
      await deleteAlertingAugmentVisSavedObj(embeddable.vis.id, modalState.monitor.id);
      notifications.toasts.addSuccess({
        title: `Association removed between the ${modalState.monitor.name} monitor and the ${title} visualization`,
        text:
          "The monitor's alerts do not automatically appear on the visualization. Refresh your dashboard to update the visualization.",
      });
      const newMonitors = monitors.filter((monitor) => monitor.id !== modalState.monitor.id);
      setAssociatedMonitors(newMonitors);
    } catch (e) {
      notifications.toasts.addDanger(
        `Failed to remove the association between the "${modalState.monitor.name}" monitor with the ${title} visualization. Failed due to ${e.message}.`
      );
    } finally {
      handleHideModal();
    }
  }, [closeFlyout, modalState]);
  const columns = useColumns({ onUnlink, onEdit });
  const emptyMsg = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body={`There are no alerting monitors associated with ${title} visualization.`}
    />
  );
  const noResultsMsg = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body="There are no alerting monitors that match the search result."
    />
  );

  const loadingMsg = <EuiEmptyPrompt body={<EuiLoadingSpinner size="l" />} />;
  const tableProps = {
    items: monitors || [],
    columns,
    search: {
      box: {
        disabled: !monitors || monitors.length === 0,
        incremental: true,
        schema: true,
      },
    },
    hasActions: true,
    pagination: true,
    sorting: true,
    message: monitors ? (monitors.length ? noResultsMsg : emptyMsg) : loadingMsg,
    loading: !monitors,
  };

  const handleHideModal = useCallback(() => {
    setModalState(undefined);
  }, []);

  return (
    <div className="associated-monitors">
      {modalState ? (
        <ConfirmUnlinkDetectorModal
          monitor={modalState.monitor}
          onHide={handleHideModal}
          onConfirm={onUnlinkMonitor}
          isListLoading={!monitors}
        />
      ) : null}
      <EuiFlyoutHeader hasBorder>
        <EuiTitle>
          <h2 id="associated-monitors__title">Associated monitors</h2>
        </EuiTitle>
      </EuiFlyoutHeader>
      {!isAssociateAllowed && (
        <EuiFlyoutHeader>
          {limitReachedCallout}
        </EuiFlyoutHeader>
      )}
      <EuiFlyoutBody>
        <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
          <EuiFlexItem>
            <EuiText>
              <h4>Visualization: {title}</h4>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButton iconType="link" isDisabled={!isAssociateAllowed} fill onClick={() => setFlyoutMode('existing')}>
              Associate a monitor
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer size="l" />
        <EuiInMemoryTable {...tableProps} />
      </EuiFlyoutBody>
    </div>
  );
};

export default AssociatedMonitors;
