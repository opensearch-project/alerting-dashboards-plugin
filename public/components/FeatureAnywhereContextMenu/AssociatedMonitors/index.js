/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import AssociatedMonitors from './AssociatedMonitors';

export default AssociatedMonitors;
