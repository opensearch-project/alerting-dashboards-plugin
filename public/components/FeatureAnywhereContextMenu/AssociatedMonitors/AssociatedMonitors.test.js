/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { notificationServiceMock } from '../../../../../../src/core/public/mocks';
import { shallow } from 'enzyme';
import AssociatedMonitors from './AssociatedMonitors';
import { setNotifications } from '../../../services';

describe('AssociatedMonitors', () => {
  test('renders', () => {
    const notificationsMock = notificationServiceMock.createSetupContract();
    setNotifications(notificationsMock);
    const wrapper = shallow(<AssociatedMonitors {...{ embeddable: { vis: { title: '' } } }} />);
    expect(wrapper).toMatchSnapshot();
  });
});
