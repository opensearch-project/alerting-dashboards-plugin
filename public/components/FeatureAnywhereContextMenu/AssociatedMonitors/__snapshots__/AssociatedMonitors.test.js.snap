// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssociatedMonitors renders 1`] = `
<div
  className="associated-monitors"
>
  <EuiFlyoutHeader
    hasBorder={true}
  >
    <EuiTitle>
      <h2
        id="associated-monitors__title"
      >
        Associated monitors
      </h2>
    </EuiTitle>
  </EuiFlyoutHeader>
  <EuiFlyoutHeader />
  <EuiFlyoutBody>
    <EuiFlexGroup
      alignItems="center"
      justifyContent="spaceBetween"
    >
      <EuiFlexItem>
        <EuiText>
          <h4>
            Visualization: 
          </h4>
        </EuiText>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <EuiButton
          fill={true}
          iconType="link"
          isDisabled={true}
          onClick={[Function]}
        >
          Associate a monitor
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
    <EuiSpacer
      size="l"
    />
    <EuiInMemoryTable
      columns={
        Array [
          Object {
            "field": "name",
            "name": "Monitor Name",
            "render": [Function],
            "sortable": true,
            "truncateText": true,
            "width": "50%",
          },
          Object {
            "field": "state",
            "name": "State",
            "render": [Function],
            "sortable": true,
            "width": "105px",
          },
          Object {
            "field": "date",
            "name": "Latest alert",
            "render": [Function],
            "sortable": true,
            "truncateText": true,
            "width": "50%",
          },
          Object {
            "actions": Array [
              Object {
                "description": "Edit monitor",
                "icon": "pencil",
                "name": "Edit monitor",
                "onClick": [Function],
                "type": "icon",
              },
              Object {
                "description": "Unlink monitor",
                "icon": "unlink",
                "name": "Unlink monitor",
                "onClick": [Function],
                "type": "icon",
              },
            ],
            "name": "Actions",
          },
        ]
      }
      hasActions={true}
      items={Array []}
      loading={true}
      message={
        <EuiEmptyPrompt
          body={
            <EuiLoadingSpinner
              size="l"
            />
          }
        />
      }
      pagination={true}
      responsive={true}
      search={
        Object {
          "box": Object {
            "disabled": true,
            "incremental": true,
            "schema": true,
          },
        }
      }
      sorting={true}
      tableLayout="fixed"
    />
  </EuiFlyoutBody>
</div>
`;
